ktor-server/ktor-server-plugins/ktor-server-websockets/jvm/test/io/ktor/tests/websocket/WebSocketTest.kt

```diff
@@ -118,593 +106,402 @@ class WebSocketTest {
-                    try {
-                        while (true) {
-                            val text = (incoming.receive() as Frame.Text).readText()
-                            received += text
-                            outgoing.send(Frame.Text(text))
-                        }
-                    } catch (e: ClosedReceiveChannelException) {
-                        // Do nothing!
-                    } catch (e: CancellationException) {
-                        // Do nothing!
-                    } catch (e: Throwable) {
-                        e.printStackTrace()
+        val received = arrayListOf<String>()
+        routing {
+            webSocket("/echo") {
+                try {
+                    while (true) {
+                        val text = receiveText()
+                        received += text
+                        send(text)
                     }
+                } catch (e: ClosedReceiveChannelException) {
+                    // Do nothing!
+                } catch (e: CancellationException) {
+                    // Do nothing!
+                } catch (e: Throwable) {
+                    e.printStackTrace()
                 }
             }
+        }
 
-            handleWebSocketConversation("/echo") { incoming, outgoing ->
-                val textMessages = listOf("HELLO", "WORLD")
-                for (msg in textMessages) {
-                    outgoing.send(Frame.Text(msg))
-                    assertEquals(msg, (incoming.receive() as Frame.Text).readText())
-                }
-                assertEquals(textMessages, received)
+        val textMessages = listOf("HELLO", "WORLD")
+        createWebSocketsClient().webSocket("/echo") {
+            for (msg in textMessages) {
+                send(msg)
+                assertEquals(msg, receiveText())
             }
         }
+
+        assertEquals(textMessages, received)
     }
 
     @Test
-    fun testFlushClosed(): Unit = withTestApplication {
-        application.install(WebSockets)
+    fun testFlushClosed() = testApplication {
+        install(WebSockets)
 
-        val session = CompletableDeferred<Unit>()
-        application.routing {
+        val session = Job()
+        routing {
             webSocket("/close/me") {
                 try {
                     close()
                     delay(1)
                     flush()
-                    session.complete(Unit)
+                    session.complete()
                 } catch (cause: Throwable) {
                     session.completeExceptionally(cause)
                     throw cause
                 }
             }
         }
 
-        handleWebSocketConversation("/close/me") { incoming, _ ->
+        var callClosedProperly = false
+        createWebSocketsClient().webSocketRaw(path = "/close/me") {
             assertTrue(incoming.receive() is Frame.Close)
             assertNull(incoming.receiveCatching().getOrNull())
+            callClosedProperly = true
```

Comment:
Speaking of missed assertion failures. `webSocketRaw` swallows all exceptions that occur in the given `block`. That's why I have to add a flag `callClosedProperly` to check later if it was changed to `true`.

It differs from `webSocket` which has no such problem, so I wonder if it is correct behavior.

Reply:
Oh, that's interesting... Maybe the block isn't connected with the parent coroutine.  Could you log a YT bug to investigate?

Reply:
Created [KTOR-7432](https://youtrack.jetbrains.com/issue/KTOR-7432/Inconsistent-behavior-when-an-exception-is-thrown-inside-webSocket-and-webSocketRaw)